### PR TITLE
Address several NeuronModelForCausalLM and TGI fixes issues

### PR DESCRIPTION
**Pull request by dacorvo**
This provides fixes and enhancements corresponding to the following issues:

- #436
- #444
- #451
- #452
- #453

### CODE REVIEW

File: optimum/neuron/generation/token_selector.py [hunk 1]
```diff
@@ -66,6 +71,9 @@ def create(
                 The model provides the internal helpers allowing to select the logits processors and stopping criterias.
             max_seq_length (`int`):
                 The maximum number of input + generated tokens for this model. It depends on the model compilation parameters.
+            stopping_criteria (`Optional[transformers.generation.StoppingCriteriaList]):
```
Comment (JingyaHuang) [inline optimum/neuron/generation/token_selector.py:74]:
```suggestion
            stopping_criteria (`Optional[transformers.generation.StoppingCriteriaList], defaults to `None`):
```

File: optimum/neuron/modeling.py [hunk 2]
```diff
@@ -747,6 +749,9 @@ def generate(
                 priority: 1) from the `generation_config.json` model file, if it exists; 2) from the model
                 configuration. Please note that unspecified parameters will inherit [`~transformers.generation.GenerationConfig`]'s
                 default values, whose documentation should be checked to parameterize generation.
+            stopping_criteria (`Optional[transformers.generation.StoppingCriteriaList]):
```
Comment (JingyaHuang) [inline optimum/neuron/modeling.py:752]:
```suggestion
            stopping_criteria (`Optional[transformers.generation.StoppingCriteriaList], defaults to `None`):
```

File: optimum/neuron/modeling_decoder.py [hunk 3]
```diff
@@ -49,6 +49,28 @@ def get_exporter(config, task):
...
+def get_available_cores() -> int:
+    """A helper to get the number of available cores.
+
+    This number depends first on the actual number of cores, then on the
+    content of the NEURON_RT_NUM_CORES and NEURON_RT_VISIBLE_CORES variables.
+    """
+    max_cores = len(os.listdir("/sys/class/neuron_device/")) * 2
+    num_cores = os.environ.get("NEURON_RT_NUM_CORES", max_cores)
+    if num_cores != max_cores:
+        num_cores = int(num_cores)
```
Comment (JingyaHuang) [inline optimum/neuron/modeling_decoder.py:62]:
```suggestion
    num_cores = int(os.environ.get("NEURON_RT_NUM_CORES", max_cores))
```

Comment (JingyaHuang) [reply optimum/neuron/modeling_decoder.py:62]: If `NEURON_RT_NUM_CORES` is higher than the max cores, would there be any informative warning or error?

**Review by JingyaHuang — APPROVED** [review]
LGTM, left some nits, thanks for the fixes!

File: optimum/neuron/modeling_decoder.py (hunk 3 above)
Comment (dacorvo) [reply optimum/neuron/modeling_decoder.py:62]: You're right: I've added some safeguards.

**Review by JingyaHuang — APPROVED** [review]

**Review by michaelbenayoun — APPROVED** [review]
LGTM, thanks for fixing these issues!

**Pull request merged**